src/lib.rs

```diff
@@ -703,12 +798,233 @@ impl Bump {
+    /// [The `_try_with` Method Suffix](#the-_try_with-method-suffix).
+    ///
+    /// ## Errors
+    ///
+    /// Iff the allocation succeeds but `f` fails, that error is forwarded by value.
+    ///
+    /// ## Panics
+    ///
+    /// Panics if reserving space for `Result<T, E>` fails.
+    ///
+    /// ## Example
+    ///
+    /// ```
+    /// let bump = bumpalo::Bump::new();
+    /// let x = bump.alloc_try_with(|| Ok("hello"))?;
+    /// assert_eq!(*x, "hello");
+    /// # Result::<_, ()>::Ok(())
+    /// ```
+    #[inline(always)]
+    #[allow(clippy::mut_from_ref)]
+    pub fn alloc_try_with<F, T, E>(&self, f: F) -> Result<&mut T, E>
+    where
+        F: FnOnce() -> Result<T, E>,
+    {
+        let rewind_footer = self.current_chunk_footer.get();
+        let rewind_ptr = unsafe { rewind_footer.as_ref() }.ptr.get();
+        let mut inner_result_ptr = NonNull::from(self.alloc_with(f));
+        let inner_result_address = inner_result_ptr.as_ptr() as usize;
+        match unsafe { inner_result_ptr.as_mut() } {
+            Ok(t) => Ok(unsafe {
+                //SAFETY:
+                // The `&mut Result<T, E>` returned by `alloc_with` may be
+                // lifetime-limited by `E`, but the derived `&mut T` still has
+                // the same validity as in `alloc_with` since the error variant
+                // is already ruled out here.
+
+                // We could conditionally truncate the allocation here, but since
+                // it grows backwards, it seems unlikely that we'd get any more
+                // than the `Result`'s discriminant this way, if anything at all.
+                &mut *(t as *mut _)
+            }),
+            Err(e) => unsafe {
+                //SAFETY:
+                // As we received `E` semantically by value from `f`, we can just
+                // copy that value here as long as we avoid a double-drop (which
+                // can't happen as any specific references to the `E`'s data in
+                // `self` are destroyed when this function returns).
+                let current_footer_p = self.current_chunk_footer.get();
+                let current_ptr = &current_footer_p.as_ref().ptr;
+                if current_ptr.get().as_ptr() as usize == inner_result_address {
+                    // We can also reuse the memory, unless `f` made any
+                    // further allocations in `self`.
+                    if current_footer_p == rewind_footer {
+                        current_ptr.set(rewind_ptr)
+                    } else {
+                        // If the current chunk changed, we can at least reset
+                        // to its start, since we know no other allocations
+                        // happened (because `current_ptr` still matches
+                        // `inner_result_address`).
+                        current_ptr.set(current_footer_p.as_ref().data)
+                    }
+                }
```

Comment:
Okay, I see why we would like to avoid `dealloc`, so we can reclaim any alignment padding added for this allocation (thanks for explaining!) However, I'd still like to reuse the `is_last_allocation` helper so that some of these pointer comparisons aren't open coded and we use "semantic" methods that make the code read smoother and better clarify intent. Additionally, I rewrote the comments "proving" soundness to convince myself that everything was above board here, and (maybe this is biased, since I wrote it, but) I think these new commetns read better:

```rust
    // If this result was the last allocation in this arena, we can reclaim its
    // space. In fact, sometimes we can do even better than simply calling
    // `dealloc` on the result pointer: we can reclaim any alignment padding we
    // might have added (which `dealloc` cannot do) if we didn't allocate a new
    // chunk for this result.
    if self.is_last_allocation(inner_result_address) {
        if current_footer_p == rewind_footer {
            // It's still the same chunk, so reset the bump pointer to its
            // original value upon entry to this method (reclaiming any
            // alignment padding we may have added).
            current_ptr.set(rewind_ptr);
        } else {
            // We allocated a new chunk for this result.
            //
            // We know the result is the only allocation in this chunk: Any
            // additional allocations since the start of this method could only
            // have happened when running the initializer function, which is
            // called *after* reserving space for this result. Therefore, since
            // we already determined via the check above that this result was
            // the last allocation, there must not have been any other
            // allocations, and this result is the only allocation in this
            // chunk.
            //
            // Because this is the only allocation in this chunk, we can reset
            // the chunk's bump finger to the start of the chunk.
            current_ptr.set(current_footer_p.as_ref().data);
        }
    }
```

Reply:
Thank you, this is better than what I wrote. I'll reflow the comments for 80 character lines (+ add a necessary cast), but will otherwise use it verbatim.